0016-pre-shared-key/0016-pre-shared-key.md

```diff
@@ -0,0 +1,218 @@
+
+The public headers come after a single status line, which must be `PSK/1.0`. The public headers are visible to all parties transmitting the packet, so they SHOULD contain only the bare minimum required for the intended parties to decrypt and validate the private headers.
+
+### Public Headers
+
+The public headers and their data are formatted as below:
+
+```http
+PSK/1.0
+Key: hmac-sha-256 fpwpAhlN588
+Header: data_everyone_can_see
+
+...
+```
+
+Everything stored in these headers is readable by all connectors between the sender and receiver. PSK payments use an HMAC of the PSK details as a fulfillment, in order to prove that the contents are unmodified. The pre-shared secret key is the key used in the HMAC.
+
+The data attached to these headers is an encrypted binary blob. This encrypted blob contains the private PSK headers.
+
+Despite the fact that connectors can read the public headers' data, connectors should not be required to do so, nor should they rely on PSK details being present on all transfers. This information is only intended for the initial sender and final receiver of a payment.
+
+The only headers in the public headers should be the ones required for the receiver to decrypt the private headers. The decryption key can be calculated by providing the data and the key to the HMAC-SHA-256 function. In the above example, the data is `fpwpAhlN588`. The key is the pre-shared secret key.
+
+| Header | Value |
+|--------|-------|
+| `Key` | _(Required)_ The HMAC algorithm, which MUST be `hmac-sha-256` (case sensitive), followed by a base64url-encoded nonce, used to generate the decryption key for the private headers. |
+| ... | _(Optional)_ Additional headers. These can be read by any connectors. |
+
+
+### Private Headers
+
+The private headers are structured like so:
+
+```http
+Expires-At: 2017-03-13T16:58:06.901Z
+Header-A: some_stuff
+Header-B: some_other_stuff
+Header-C: some_more_stuff
+Date: 2017-03-13T16:57:56.901Z
+Content-Type: application/json
+
+{"foo":"bar"}
+```
+
+Only the intended sender and receiver can decrypt and read the private headers and their data. The private headers contains an arbitrary message body preceded by the following headers:
+
+| Header | Value |
+|--------|-------|
+| `Expires-At` | _(Optional)_ An ISO8601 UTC timestamp, after which this PSK payment has expired. |
+| `Content-Type` | _(Optional)_ The [media type](http://www.iana.org/assignments/media-types/media-types.xhtml) of the data in the message body |
+| ... | _(Optional)_ Additional arbitrary headers. Where appropriate, it is RECOMMENDED to match the headers defined for HTTP messages ([RFC7230](https://tools.ietf.org/html/rfc7230#section-3.2)) |
+
+# Pseudocode
+
```

Comment:
Add "The following pseudocode outlines the generation procedure for the keys involved in PSK"